### ISSUE TITLE
Aliasing a blob with custom type in C++ generates compilation errors

### ISSUE THREAD

**Issue by gravikovich** [microsoft/bond#999]
I'm trying to perform aliasing of an encoded image using a blob (MSVC compiler)

Given a bond definition of `images_structs.bond`: 
```
namespace Tests

using Image = blob;

struct ImageStruct
{
    0: Image ImageField;
}
```

and codegen of:
```
gbc.exe c++ --header="..." --using Image=MyImageType -o=.  images_structs.bond 
```

and alias converters of:
```cpp
namespace bond
{
    template <> struct
        aliased_type<MyImageType>
    {
        typedef bond::blob type;
    };
}


void set_aliased_value(MyImageType& var, const bond::blob& value){...}

bond::blob get_aliased_value(const MyImageType& value){...}

```
I got several compilation error:
* operator== wasn't defined on MyImageType. 
   * I defined it. But then:
* core\detail\omit_default.h(43): error C2679: binary '==': no operator found which takes a right-hand operand of type 'bond::blob' (or there is no acceptable conversion)
  * After some tracing I defined:
```
namespace bond
{
    namespace detail
    {
        bool is_default(const MyImageType& /*value*/, const Metadata& /*metadata*/)
        {
            return false;
        }
    }
}
```

Is this the expected behavior? Is it supported?
The equivalent code in C# worked just fine, without any 'workarounds'...

**Comment by gravikovich** [comment]
Continuing the saga - the above patches worked for serialization, for deserialization I get the error:
```
core\value.h(316): error C2664: 'void bond::CompactBinaryReader<bond::InputBuffer>::Read(bond::blob &,uint32_t)': cannot convert argument 1 from 'bond::aliased_type<MyImageType>::type' to 'int8_t &'
```

**Comment by chwarr** [comment]
Take a look at the [multiprecision example](https://github.com/microsoft/bond/tree/master/examples/cpp/core/multiprecision). blobs are container types, so aliasing them works differently than scalars.

**Comment by gravikovich** [comment]
@chwarr Thanks for the prompt response.
Since I want to perform an encoding of the image (i.e. JPG/PNG/...) the desired and efficient way would be to reuse the blob buffer and encode to/from it directly, not element by element.
In C# I could just alias `ArraySegment<byte>`:

```

        public static MyImage Convert(ArraySegment<byte> value, BondedImage unused)
        {
              //Decode directly from the value buffer, create my image type
        }

        public static ArraySegment<byte> Convert(MyImage value, ArraySegment<byte> unused)
        {
            // Create the required buffer, encode, and wrap in ArraySegment
        } 

```

This was very clean and efficient. Is there any way to do this in C++?  
Otherwise I would need to create a temporary buffer, encode to it (when should I do that? In the constructor of the enumerator?) and provide iteration over it for the library (which if I understand correctly will copy it into the blob). For the decoding - I would need to know when the writing into that temp buffer was completed somehow (destructor? when the iteration reaches the end?) decode it into the stored reference to MyImage class.

**Comment by chwarr** [comment]
As you've noticed, the C# implementation tends to special case blobs as something apart from a list/vector container and apart from a scalar. C++ treats blobs like list/vector containers the vast majority of the time.

> Otherwise I would need to create a temporary buffer, encode to it (when should I do that? In the constructor of the enumerator?)

Yes, the constructor of the _const_ enumerator is a reasonable place to do the encoding. (The _const_ enumerator will be used for reading as bytes, while the _non-const_ enumerator will be used for populating with bytes.) Depending on your call patterns, you may want to cache the encoded representation (e.g., encode on-demand, store it in the image class and then give a `shared_ptr` to each const enumerator to iterate over, invalidate upon mutations). A little bit of instrumentation will help you make the right trade off here.

> and provide iteration over it for the library

Correct: the const enumerator would iterate over the encoded bytes.

> which if I understand correctly will copy it into the blob

In practical use, no, there won't be a blob created. It will instead [copy the bytes from the image's encoded form into whatever output stream is being used at serialization time](https://github.com/microsoft/bond/blob/8d0842280856f357664d6ea9d3d010497444e8ab/cpp/inc/bond/core/transforms.h#L263-L276). There may be cases when a blob is created, depending on exactly which transforms are being applied.

> For the decoding - I would need to know when the writing into that temp buffer was completed somehow (destructor? when the iteration reaches the end?) decode it into the stored reference to MyImage class.

Yes, this is also similar to the serialization case. The destructor for the _non-const_ enumerator would be a good place to decode the bytes. You could use something like `std::variant<std::vector<char>, UnderlyingImageRepresentation>>` so that you don't have to have space in your image class for both the buffer and the image.

Hope this helps.

**Comment by gravikovich** [comment]
Thanks, I will try and update. Side note - is there really a need to implement `operator==` on aliased types? The codegen generates the comparison operator automatically.

**Comment by chwarr** [comment]
> is there really a need to implement operator== on aliased types? The codegen generates the comparison operator automatically.

In C++, Bond structs and their fields have value semantics. If two objects with different identities have the same data in them, they should compare equal to each other. (E.g., deserializing the same Bond payload into two different objects results in objects that are equal to each other.) If the implementation of the alias type doesn't have value semantics for `operator==`, you'll experience odd behavior like the Bond generated structs' `operator==` not working as expected. Codegen [defers to the types' `operator==` in its implementation](https://github.com/microsoft/bond/blob/29c65c828a3efc641f8d9abbad05068ed89d677a/compiler/tests/generated/aliases_types.h#L57-L61).

Serialization doesn't call the alias type's `operator==`. It looks at [the wire value for scalar fields](https://github.com/microsoft/bond/blob/29c65c828a3efc641f8d9abbad05068ed89d677a/cpp/inc/bond/core/detail/omit_default.h#L37-L44) to determine whether the field can be omitted or not. For containers, it checks whether [the container is empty](https://github.com/microsoft/bond/blob/29c65c828a3efc641f8d9abbad05068ed89d677a/cpp/inc/bond/core/detail/omit_default.h#L68-L75). So, you don't need to worry about the performance of your `operator==` implementation for serialization/deserialization.

Hope this helps.

**Comment by gravikovich** [comment]
Thanks, I have succeeded aliasing the blob using your guidelines.
An unexpected issues I encountered was that I couldn't do the encoding in the constructor of the `const_enumerator` since the library called `container_size` to check for default values before it's creating the const_enumerator. Also the library encodes the container size prior to the `const_enumerator` serialization. So I ended up with code like:
```cpp
    inline
    std::uint32_t container_size(const MyImage& img)
    {
        if (!img.empty() && img.encodedBuffer.empty()) //not yet encoded
        {
            img.encodeImage();
        }
        return img.encodedBuffer.size();
    }
```
Which is a bit confusing, but I can live with it. My 2 cents to make this a bit easier for the future:
* I would suggest to allow the user to specialize `is_default` to allow the user not having to deal with comparison unless required.
* A single "concept" specialization class that will server as a utility holder for all aliasing needs, instead of several traits in bond namespace, several free methods, several specializations classes. Something like:

```cpp
namespace bond
{
template<>
struct serialization_alias<MyClass> : list_alias<std::int8_t>
{
   serialization_alias(const MyClass&) {...} //called before encoding of the class starts
   serialization_alias(MyClass&) {...} //called before decoding of the class begins

   std::uint32_t size() const { ...}
   void resize(std::uint32_t) {...}

   bool more() const;
   std::int8_t next() const;
   std::int8_t& next();
};
}
```
This will also allow storing the intermediate buffers in this class and not in the user defined model.
Anyhow - thanks a lot for the help, highly appreciated.
